tools/prow-auto-bumper/file.go

```diff
@@ -0,0 +1,122 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+func cdToRootDir() error {
+	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
+	output, err := cmd.Output()
+	if err != nil {
+		return err
+	}
+	d := strings.TrimSpace(string(output))
+	log.Printf("Changing working directory to %s...", d)
+	return os.Chdir(d)
+}
+
+func call(cmd string, args ...string) error {
+	c := exec.Command(cmd, args...)
+	c.Stdout = os.Stdout
+	c.Stderr = os.Stderr
+	return c.Run()
+}
+
+// update all tags in a byte slice
+func updateAllTags(pv *PRVersions, content []byte, imageFilter *regexp.Regexp) ([]byte, string, []string) {
+	var msg string
+	var errMsgs []string
+	indexes := imageFilter.FindAllSubmatchIndex(content, -1)
+	// Not finding any images is not an error.
+	if indexes == nil {
+		return content, msg, errMsgs
+	}
+
+	var res string
+	lastIndex := 0
+	for _, m := range indexes {
+		res += string(content[lastIndex : m[imageImagePart*2+1]+1])
+		image := string(content[m[imageImagePart*2]:m[imageImagePart*2+1]])
+		tag := string(content[m[imageTagPart*2]:m[imageTagPart*2+1]])
+		lastIndex = m[1]
+
+		iv := pv.getIndex(image, tag)
+		if "" != pv.images[image][iv].newVersion {
+			res += pv.images[image][iv].newVersion
+			msg += fmt.Sprintf("\nImage: %s\nOld Tag: %s\nNew Tag: %s", image, tag, pv.images[image][iv].newVersion)
+		} else {
+			errMsg := fmt.Sprintf("Cannot find version for image: '%s:%s'.\n", image, tag)
+			log.Println(errMsg)
+			errMsgs = append(errMsgs, errMsg)
+			res += tag
+		}
+	}
+	res += string(content[lastIndex:])
+
+	return []byte(res), msg, errMsgs
+}
+
+// updateFile updates a file in place.
+func updateFile(pv *PRVersions, filename string, imageFilter *regexp.Regexp, dryrun bool) ([]string, error) {
+	var errMsgs []string
+	content, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return errMsgs, fmt.Errorf("failed to read %s: %v", filename, err)
+	}
+
+	newContent, msg, errMsgs := updateAllTags(pv, content, imageFilter)
+	if err := run(
+		fmt.Sprintf("Update file '%s':%s", filename, msg),
+		func() error {
+			return ioutil.WriteFile(filename, newContent, 0644)
+		},
+		dryrun); err != nil {
+		return errMsgs, fmt.Errorf("failed to write %s: %v", filename, err)
+	}
+	return errMsgs, nil
+}
+
+func updateAllFiles(pv *PRVersions, fileFilters []*regexp.Regexp, imageFilter *regexp.Regexp,
+	dryrun bool) ([]string, error) {
+	var errMsgs []string
```

Comment:
Do we indeed need to return this `errMsgs`? Seems it's empty if `err` is `nil`, and it's not being used if `err` is not `nil`.

Reply:
It's not used here, the purpose of errMsgs is to be displayed in Version update PR, so approvers get first hand idea